### PR TITLE
Impove(enter): enter leader mds directly without id option

### LINKED ISSUE #324: curveadm enter command can directly enter the leader container
现在我们想要进入leader容器的话，需要使用curveadm status查看leader容器的id，然后使用curveadm enter <id>进入，现在我们或许可以使用curveadm enter不用输入id的话就默认进入leader容器。

## Registration procedure
Step 1. Add a comment when you sign up.The purpose of comments is for us to understand you and determine whether you have the basic ability to complete this issue. 

- If you are a student, please introduce your school, major, programming language, technical fields of interest, what you are good at, and why you chose this issue. 
- If you are already working, please introduce your company, career, work content, technical fields of interest, and why you chose this issue.

Step 2. Use WeChat to add the Curve community assistant "Curve" as a WeChat friend, tell your GitHub ID, and send the Issue link of your choice. The assistant will invite you to communicate with the mentor  by WeChat.

Step 3. In order to better complete each development task, the mentor will select the developer who best matches the issue based on everyone's comment on GitHub and communication status by WeChat. If you are not selected, please select another issue.

## Tips
- Before developing, please read the dev guideline( https://ask.opencurve.io/t/topic/201) carefully and communicate with your mentor.
- Please submit a draft PR before 11.29, and the mentor will review it within 2-5 working days and may guide you to modify the PR based on the actual situation.
- If you successfully merge into PR before 12.27 and you will receive gifts carefully prepared by the Curve community.

### PR DESCRIPTION
**Pull request by LYPWYT**
fix #324 
Modify the enter command to directly enter the leader container without any parameter

### CODE REVIEW

File: internal/task/task/common/service_status.go [hunk 1]
```diff
@@ -306,3 +334,50 @@ func NewGetServiceStatusTask(curveadm *cli.CurveAdm, dc *topology.DeployConfig)
...
+	var status string
+	var isLeader bool
+	t.AddStep(&step.ListContainers{
+		ShowAll:     true,
+		Format:      `"{{.Status}}"`,
+		Filter:      fmt.Sprintf("id=%s", containerId),
+		Out:         &status,
+		ExecOptions: curveadm.ExecOptions(),
+	})
+	t.AddStep(&step.Lambda{
```
Comment (caoxianfei1) [inline internal/task/task/common/service_status.go:370]: Here we need to check container is Exist.

File: internal/task/task/common/service_status.go [hunk 2]
```diff
@@ -306,3 +334,50 @@ func NewGetServiceStatusTask(curveadm *cli.CurveAdm, dc *topology.DeployConfig)
...
+	subname := fmt.Sprintf("host=%s role=%s containerId=%s",
+		dc.GetHost(), dc.GetRole(), tui.TrimContainerId(containerId))
+	t := task.NewTask("Enter Leader container", subname, hc.GetSSHConfig())
+
+	// add step to task
+	var status string
+	var isLeader bool
+	t.AddStep(&step.ListContainers{
+		ShowAll:     true,
+		Format:      `"{{.Status}}"`,
```
Comment (caoxianfei1) [inline internal/task/task/common/service_status.go:365]: {{.ID}} is better? our goal is check container whether exist and not to get  status of service container.

Comment (caoxianfei1) [reply internal/task/task/common/service_status.go:365]: The code is already exist.
```shell
	t.AddStep(&step.ListContainers{
		ShowAll:     true,
		Format:      `"{{.ID}}"`,
		Filter:      fmt.Sprintf("id=%s", containerId),
		Out:         &out,
		ExecOptions: curveadm.ExecOptions(),
	})
	t.AddStep(&step.Lambda{
		Lambda: CheckContainerExist(host, role, containerId, &out),
	})
```

File: internal/task/task/common/service_status.go [hunk 3]
```diff
@@ -306,3 +334,50 @@ func NewGetServiceStatusTask(curveadm *cli.CurveAdm, dc *topology.DeployConfig)
...
+		Lambda: TrimContainerStatus(&status),
+	})
+	t.AddStep(&step2GetLeader{
+		dc:          dc,
+		containerId: containerId,
+		status:      &status,
+		isLeader:    &isLeader,
+		execOptions: curveadm.ExecOptions(),
+	})
+	t.AddStep(&step2FormatMdsLeaderId{
```
Comment (caoxianfei1) [inline internal/task/task/common/service_status.go]: step2SetMdsLeader

File: internal/task/task/common/service_status.go [hunk 4]
```diff
@@ -109,6 +115,21 @@ func setServiceStatus(memStorage *utils.SafeMap, id string, status ServiceStatus
...
+			return nil
+		}
+		if IsLeader {
+			m = id
+		}
+		kv.Set(comm.MDS_LEADER_ID, m)
+		return nil
+	})
+}
+
```
Comment (caoxianfei1) [inline internal/task/task/common/service_status.go]: There is no need to extract to a function.

File: cli/command/enter.go [hunk 5]
```diff
@@ -57,32 +62,85 @@ func NewEnterCommand(curveadm *cli.CurveAdm) *cobra.Command {
...
 		return err
 	}
+	var containerId string
+	var dc *topology.DeployConfig
+	Id := options.id
 
-	// 2) filter service
+	// 2) If no id parameter, execute the following
+	if Id == "" {
+		// generate get status playbook
```
Comment (caoxianfei1) [inline cli/command/enter.go]: comment is incorrect. Delete it or  modify.

File: cli/command/enter.go [hunk 6]
```diff
@@ -57,32 +62,85 @@ func NewEnterCommand(curveadm *cli.CurveAdm) *cobra.Command {
...
+		statusForLeaderOptions := statusOptions{id: "*", role: ROLE_MDS, host: "*"}
```
Comment (caoxianfei1) [inline cli/command/enter.go]: update the name `statusForLeaderOptions`

File: cli/command/enter.go [hunk 7]
```diff
@@ -57,32 +62,85 @@ func NewEnterCommand(curveadm *cli.CurveAdm) *cobra.Command {
...
+	steps := []int{playbook.GET_MDS_LEADER}
+	pb := playbook.NewPlaybook(curveadm)
+	for _, step := range steps {
+		pb.AddStep(&playbook.PlaybookStep{
+			Type:    step,
+			Configs: dcs,
+			ExecOptions: playbook.ExecOptions{
+				//Concurrency:   10,
+				SilentSubBar:  true,
+				SilentMainBar: step == playbook.INIT_SERVIE_STATUS,
```
Comment (caoxianfei1) [inline cli/command/enter.go]: DELETE the `playbook.INIT_SERVIE_STATUS` and not used.

Comment (caoxianfei1) [reply cli/command/enter.go]: Set true directly is ok.

File: cli/command/enter.go [hunk 8]
```diff
@@ -57,32 +62,85 @@ func NewEnterCommand(curveadm *cli.CurveAdm) *cobra.Command {
...
+	dcs = curveadm.FilterDeployConfig(dcs, topology.FilterOption{
+		Id:   options.id,
+		Role: options.role,
+		Host: options.host,
+	})
+	if len(dcs) == 0 {
+		return nil, errno.ERR_NO_SERVICES_MATCHED
+	}
+
+	steps := []int{playbook.GET_MDS_LEADER}
```
Comment (caoxianfei1) [inline cli/command/enter.go]: Define a variable like other command.
```shell
var (
   AttachMdsLeaderContainer = []int{playbook.ATTACH_MDS_LEADER_CONTAINER}
)
```

**Review by caoxianfei1 — APPROVED** [review]
LGTM!

File: cli/command/enter.go [hunk 9]
```diff
@@ -57,32 +66,83 @@ func NewEnterCommand(curveadm *cli.CurveAdm) *cobra.Command {
...
 	if err != nil {
 		return err
 	}
+	var containerId string
+	var dc *topology.DeployConfig
+	id := options.id
 
-	// 2) filter service
+	// 2) If no id options, get Mds leader id
+	if id == "" {
```
Comment (Wine93) [inline cli/command/enter.go]: suggestion: use a function to do this.

Comment (LYPWYT) [reply cli/command/enter.go]: Thanks for comments! Finish!

File: cli/command/enter.go [hunk 10]
```diff
@@ -57,32 +66,95 @@ func NewEnterCommand(curveadm *cli.CurveAdm) *cobra.Command {
...
+	options statusOptions) (*playbook.Playbook, error) {
+	dcs = curveadm.FilterDeployConfig(dcs, topology.FilterOption{
+		Id:   options.id,
+		Role: options.role,
+		Host: options.host,
+	})
+	if len(dcs) == 0 {
+		return nil, errno.ERR_NO_SERVICES_MATCHED
+	}
+
```
Comment (caoxianfei1) [inline cli/command/enter.go]: unnecessary

File: cli/command/enter.go [hunk 11]
```diff
@@ -57,32 +66,95 @@ func NewEnterCommand(curveadm *cli.CurveAdm) *cobra.Command {
...
+				SilentMainBar: true,
+				SkipError:     true,
+			},
+		})
+	}
+	return pb, nil
+}
+
+func getLeaderOrRandomId(curveadm *cli.CurveAdm, dcs []*topology.DeployConfig, id string) (string, error) {
+	statusForMdsLeaderOptions := statusOptions{id: "*", role: "*", host: "*"}
```
Comment (caoxianfei1) [inline cli/command/enter.go]: unnecessary.

File: cli/command/enter.go [hunk 12]
```diff
@@ -57,32 +66,95 @@ func NewEnterCommand(curveadm *cli.CurveAdm) *cobra.Command {
...
+	id := options.id
 
-	// 2) filter service
+	// 2) If no id options, get leader or random container id
+	if id == "" {
+		id, err = getLeaderOrRandomId(curveadm, dcs, id)
+		if err != nil {
+			return err
+		}
+	}
```
Comment (caoxianfei1) [inline cli/command/enter.go]: Use a uniform function to get the final id. such as checkOrGetId.

File: internal/task/task/common/service_status.go [hunk 13]
```diff
@@ -306,3 +335,51 @@ func NewGetServiceStatusTask(curveadm *cli.CurveAdm, dc *topology.DeployConfig)
...
+	// new task
+	subname := fmt.Sprintf("host=%s role=%s containerId=%s",
+		dc.GetHost(), dc.GetRole(), tui.TrimContainerId(containerId))
+	t := task.NewTask("Enter Leader container", subname, hc.GetSSHConfig())
+
+	// add step to task
+	var status string
+	var isLeader bool
+	t.AddStep(&step.ListContainers{
+		ShowAll:     true,
```
Comment (caoxianfei1) [inline internal/task/task/common/service_status.go]: -all parameters will show the exited container.

File: cli/command/enter.go [hunk 14]
```diff
@@ -57,32 +66,87 @@ func NewEnterCommand(curveadm *cli.CurveAdm) *cobra.Command {
...
+	id = value.([2]string)[0]
+	id_random := value.([2]string)[1]
+	if id == "" && id_random == "" {
+		return "", errno.ERR_NO_LEADER_OR_RANDOM_CONTAINER_FOUND
+	} else if id != "" {
+		return id, nil
+	} else {
+		return id_random, nil
+	}
+}
```
Comment (caoxianfei1) [inline cli/command/enter.go]: Ditto.

File: internal/task/task/common/service_status.go [hunk 15]
```diff
@@ -218,6 +225,28 @@ func (s *step2FormatServiceStatus) Execute(ctx *context.Context) error {
...
+		m := [2]string{"", ""}
+		m[1] = id
+		v := kv.Get(comm.LEADER_OR_RANDOM_ID)
+		if v != nil && v.([2]string)[0] != "" {
+			return nil
+		}
+		if IsLeader {
+			m[0] = id
+		}
+		kv.Set(comm.LEADER_OR_RANDOM_ID, m)
```
Comment (caoxianfei1) [inline internal/task/task/common/service_status.go]: Only set the final id is ok. not an array.

**Review by Wine93 — APPROVED** [review]

**Pull request merged**